widget/entry_cursor_anim.go

```diff
@@ -0,0 +1,156 @@
+package widget
+
+import (
+	"image/color"
+	"sync"
+	"time"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/canvas"
+	"fyne.io/fyne/v2/theme"
+)
+
+type safeCounter struct {
+	mu  sync.RWMutex
+	cnt int
+}
+
+func (c *safeCounter) Inc(n int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.cnt += n
+}
+
+func (c *safeCounter) Value() int {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.cnt
+}
+
+func (c *safeCounter) Reset() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.cnt = 0
+}
+
+const cursorStopTimex10ms = 35 // stop time in multiple of 10 ms
+
+type cursorState int
+
+const (
+	cursorStateRunning cursorState = iota
+	cursorStateTemporarilyStopped
+	cursorStateStopped
+)
+
+type entryCursorAnimation struct {
+	mu       *sync.RWMutex
+	counter  *safeCounter
+	inverted bool
+	state    cursorState
+	cursor   *canvas.Rectangle
+	anim     *fyne.Animation
+	sleepFn  func(time.Duration) // added for tests, do not change it outside of tests!!
+}
+
+func newEntryCursorAnimation(cursor *canvas.Rectangle) *entryCursorAnimation {
+	a := &entryCursorAnimation{mu: &sync.RWMutex{}, cursor: cursor, counter: &safeCounter{}}
+	a.inverted = false
+	a.state = cursorStateStopped
+	a.sleepFn = time.Sleep
+	return a
+}
+
+func (a *entryCursorAnimation) createAnim(inverted bool) *fyne.Animation {
+	cursorOpaque := theme.PrimaryColor()
+	r, g, b, _ := theme.PrimaryColor().RGBA()
+	cursorDim := color.NRGBA{R: uint8(r), G: uint8(g), B: uint8(b), A: 0x16}
+	a.inverted = inverted
+	start, end := color.Color(cursorDim), color.Color(cursorOpaque)
+	if inverted {
+		start, end = end, start
+	}
+	anim := canvas.NewColorRGBAAnimation(start, end, time.Second/2, func(c color.Color) {
+		a.cursor.FillColor = c
+		a.cursor.Refresh()
+	})
+	anim.RepeatCount = fyne.AnimationRepeatForever
+	anim.AutoReverse = true
+	return anim
+}
+
+// Start starts cursor animation.
+func (a *entryCursorAnimation) Start() {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	// anim should be nil and current state should be stopped
+	if a.anim != nil || a.state != cursorStateStopped {
+		return
+	}
+	a.anim = a.createAnim(false)
+	a.state = cursorStateRunning
+	go func() {
+		defer func() {
+			a.mu.Lock()
+			a.state = cursorStateStopped
+			a.mu.Unlock()
+		}()
+		for {
```

Comment:
Looking at the counter and the loop and I wondered if an alternative approach might be easier to read (and less background processing)... this is just thinking out loud.

Could you wait until you suspect the cursor would wake up (i.e. simple delay), then at that event see if the time has lapsed, and if not (because a new delay was added) then wait the difference between the first delay and the latter, eventually seeing there is no additional delay and return to the running animation? 

This would also mean that the animation wakes up after the specified time, without needing to check every 10ms

Reply:
Yes, that was the first thing I had in mind, but the problem is the reset part, if we do not check every "x" time then we will introduce some delays in the animation, or maybe I am wrong?

Reply:
Perhaps I don't follow but my understanding was that it would always pause at least the delay time, which may get longer if the user keeps typing. So what is the reset event that we need to check sooner?

Reply:
Hmm ok, I will implement the logic with a ticker instead maybe it is more clear with it?

Reply:
I would have thought it would be. Worth exploring.
If it is cleaner without let me know and I will do a more thorough review of this code

Reply:
@andydotxyz  Ok, I have updated the code using now time.Ticker. I am just worry if calling ticker.Reset() so quickly would lead to a bad performance.

Reply:
However, now I realized that time.Ticker does not have Reset() method in Go 1.12 😞

Reply:
@andydotxyz  I have rewritten the solution with time.Timer instead of time.Ticker to ensure backwards compatibility with Go. Tests are ok now! :)